### PR TITLE
ts/parallelaxis

### PR DESCRIPTION
**Pull request by bre1470**
Migrated ParallelAxis to ES6 pattern.

### CODE REVIEW

File: ts/modules/parallel-coordinates.src.ts [hunk 1]
```diff
@@ -315,114 +485,6 @@ extend(ChartProto, /** @lends Highcharts.Chart.prototype */ {
 });
 
 
-// On update, keep parallelPosition.
-Axis.keepProps.push('parallelPosition');
```
Comment (pawelfus) [inline ts/modules/parallel-coordinates.src.ts]: I think this is necessary?

Comment (bre1470) [reply ts/modules/parallel-coordinates.src.ts]: If that is the case, do I have to add the `parallel` property here?

Comment (bre1470) [reply ts/modules/parallel-coordinates.src.ts]: Please review.

Comment (pawelfus) [reply ts/modules/parallel-coordinates.src.ts]: In general, when Axis is updated, it's recreated from scratch. However, some properties are stored (from array `keepProps`) and assigned back.

I don't know if that is still necessary for `parallelPosition` property (or the new one `parallel.position`). Here is the case: https://jsfiddle.net/BlackLabel/c0m8L7d6/3/ - after 1s, updated second yAxis without issues. Here is current build: https://jsfiddle.net/BlackLabel/c0m8L7d6/2/ - btw, first axis is missing.

Comment (pawelfus) [reply ts/modules/parallel-coordinates.src.ts]: Note that other compositions may not need that property at all.

Comment (bre1470) [reply ts/modules/parallel-coordinates.src.ts]: Okay, I fixed it and add compositions to keepProps only, if it makes sense. You fiddle works with the last commit: https://jsfiddle.net/1zu08j4t/

File: ts/parts/DateTimeAxis.ts [hunk 2]
```diff
@@ -211,6 +211,8 @@ class DateTimeAxis {
      */
     public static compose(AxisClass: typeof Axis): void {
 
+        AxisClass.keepProps.push('dateTime');
```
Comment (pawelfus) [inline ts/parts/DateTimeAxis.ts]: So `dateTime` does need to be in `keepProps`? Why?

Comment (bre1470) [reply ts/parts/DateTimeAxis.ts]: For performance reasons, as we skip creating a new instance during re-init of the axis.

File: ts/parts/NavigatorAxis.ts [hunk 3]
```diff
@@ -155,11 +155,16 @@ class NavigatorAxis {
      */
     public static compose(AxisClass: typeof Axis): void {
 
+        AxisClass.keepProps.push('navigatorAxis');
```
Comment (pawelfus) [inline ts/parts/NavigatorAxis.ts]: I guess this one makes sense because of `Axis.navigatorAxis.previousZoom`?

File: ts/parts/OrdinalAxis.ts [hunk 4]
```diff
@@ -355,11 +355,12 @@ class OrdinalAxis {
         SeriesClass: typeof Series
     ): void {
 
+        AxisClass.keepProps.push('ordinal');
```
Comment (pawelfus) [inline ts/parts/OrdinalAxis.ts]: I don't see a reason to keep `ordinal` prop.

File: ts/parts/LogarithmicAxis.ts [hunk 5]
```diff
@@ -215,7 +215,7 @@ class LogarithmicAxis {
      */
     public static compose(AxisClass: typeof Axis): void {
 
-        /* eslint-disable no-invalid-this */
+        AxisClass.keepProps.push('logarithmic');
```
Comment (pawelfus) [inline ts/parts/LogarithmicAxis.ts]: I think this one actually makes sense 🤔I see there's code to be executed when switching from `log` to `linear` type. Nice catch! 👍

File: ts/parts/OrdinalAxis.ts (hunk 4 above)
Comment (bre1470) [reply ts/parts/OrdinalAxis.ts]: So we instantiate it only ones per axis instance and not with every axis init: https://github.com/highcharts/highcharts/blob/5133f3e50d90fac52261571ac43fd0c76a076de5/ts/parts/OrdinalAxis.ts#L939 Condition is missing for that though.

File: ts/parts/LogarithmicAxis.ts (hunk 5 above)
Comment (bre1470) [reply ts/parts/LogarithmicAxis.ts]: I disagree with you. Saving constructor calls is sensible, too.

File: ts/parts/DateTimeAxis.ts (hunk 2 above)
Comment (pawelfus) [reply ts/parts/DateTimeAxis.ts]: Just to explain: It's similar to `logarithmic`, but the only action on `destroy` is to delete `axis.dateTime.axis`.

File: ts/parts/LogarithmicAxis.ts (hunk 5 above)
Comment (pawelfus) [reply ts/parts/LogarithmicAxis.ts]: But the code is clearing props only on `axis.linear` object. I'm not so convinced anymore.. :)

Comment (pawelfus) [reply ts/parts/LogarithmicAxis.ts]: Above comment was added before your loaded for me, I'm commenting myself comment above. Saying that, I don't understand what do you mean.

File: ts/parts/DateTimeAxis.ts (hunk 2 above)
Comment (pawelfus) [reply ts/parts/DateTimeAxis.ts]: > For performance reasons, as we skip creating a new instance during re-init of the axis.

Yes, that makes sense! 👍

**Comment by bre1470** [comment]
I wonder, if we should introduce a better way than keepProps for complex objects. How about a `keepMe` property, that `Axis.destroy` checks on its properties? For example `Axis.ordinal.keepMe` would be true and so it would skip the keepProps check.

File: ts/parts/LogarithmicAxis.ts (hunk 5 above)
Comment (bre1470) [reply ts/parts/LogarithmicAxis.ts]: I removed the destroy over design. :)
I mean this part here: https://github.com/highcharts/highcharts/blob/5133f3e50d90fac52261571ac43fd0c76a076de5/ts/parts/LogarithmicAxis.ts#L242-L244

File: ts/parts/OrdinalAxis.ts (hunk 4 above)
Comment (bre1470) [reply ts/parts/OrdinalAxis.ts]: I added the condition for ordinal composition.

**Comment by pawelfus** [comment]
> For example `Axis.ordinal.keepMe` would be true and so it would skip the `keepProps` check.

FYI: We do that in Series: `Series.propertyName.survive` (for SVG elements), but it's not much different from `keepProps` check.

**Comment by bre1470** [comment]
@pawelfus I addressed all you concerns (I hope). Please review ones more.

**Comment by bre1470** [comment]
@TorsteinHonsi Should we introduce a sub-property `survive` for the `Axis.destroy` loop?

**Comment by TorsteinHonsi** [comment]
> Should we introduce a sub-property survive for the Axis.destroy loop?

Can't do that, these properties may be any type. `extKey` for example is a string, `userMin` is `number|undefined`.

The `keepProps` concept is light-weight and has been working well (pragmatic in other words). I don't quite see a reason for changing it?

**Comment by bre1470** [comment]
@TorsteinHonsi For me this is isolated: https://github.com/highcharts/highcharts/blob/d17c85e73866e46ac9cb99f9969eda811d23ea5f/ts/modules/parallel-coordinates.src.ts#L166-L237 Or did you meant something else?

**Comment by TorsteinHonsi** [comment]
Yes, I meant these additional properties on the Axis and Chart:

```js
        interface Axis {
            parallelPosition?: ParallelAxis['parallelPosition'];
            /** @requires modules/parallel-coordinates */
            setParallelPosition(
                axisPosition: Array<string>,
                options: AxisOptions
            ): void;
        }
        interface Chart {
            hasParallelCoordinates?: ParallelChart['hasParallelCoordinates'];
            parallelInfo?: ParallelChart['parallelInfo'];
            /** @requires modules/parallel-coordinates */
            setParallelInfo(options: Options): void;
        }
```

**Comment by bre1470** [comment]
We did not discussed the Chart migrasjon yet, so I leave it everywhere out.

**Comment by bre1470** [comment]
I will create new PRs. This becomes confusing because of the keepProps side stuff.

**Pull request closed without merging**